src/components/views/settings/ThemeChoicePanel.tsx

```diff
@@ -1,285 +1,341 @@
+        </Root>
+    );
 }
 
-interface IState extends IThemeState {
-    customThemeUrl: string;
-    customThemeMessage: CustomThemeMessage;
+/**
+ * Component to select the theme
+ */
+interface ThemeSelectorProps {
+    /* The current theme */
+    theme: string;
+    /* The theme can't be selected */
+    disabled: boolean;
 }
 
-export default class ThemeChoicePanel extends React.Component<IProps, IState> {
-    private themeTimer?: number;
+/**
+ * Component to select the theme
+ */
+function ThemeSelectors({ theme, disabled }: ThemeSelectorProps): JSX.Element {
+    const themes = useThemes();
+
+    return (
+        <Root
+            className="mx_ThemeChoicePanel_ThemeSelectors"
+            onChange={async (evt) => {
+                // We don't have any file in the form, we can cast it as string safely
+                const newTheme = new FormData(evt.currentTarget).get("themeSelector") as string | null;
+
+                // Do nothing if the same theme is selected
+                if (!newTheme || theme === newTheme) return;
+
+                // doing getValue in the .catch will still return the value we failed to set,
+                SettingsStore.setValue("theme", null, SettingLevel.DEVICE, newTheme).catch(() => {
+                    dis.dispatch<RecheckThemePayload>({ action: Action.RecheckTheme });
+                });
 
-    public constructor(props: IProps) {
-        super(props);
+                // The settings watcher doesn't fire until the echo comes back from the
+                // server, so to make the theme change immediately we need to manually
+                // do the dispatch now
+                // XXX: The local echoed value appears to be unreliable, in particular
+                // when settings custom themes(!) so adding forceTheme to override
+                // the value from settings.
+                dis.dispatch<RecheckThemePayload>({ action: Action.RecheckTheme, forceTheme: newTheme });
+            }}
+        >
+            {themes.map((_theme) => {
+                const isChecked = theme === _theme.id;
+                return (
+                    <InlineField
+                        className={classNames("mx_ThemeChoicePanel_themeSelector", {
+                            [`mx_ThemeChoicePanel_themeSelector_enabled`]: !disabled && theme === _theme.id,
+                            [`mx_ThemeChoicePanel_themeSelector_disabled`]: disabled,
+                            // We need to force the compound theme to be light or dark
+                            // The theme selection doesn't depend on the current theme
+                            // For example when the light theme is used, the dark theme selector should be dark
+                            "cpd-theme-light": !_theme.isDark,
+                            "cpd-theme-dark": _theme.isDark,
+                        })}
+                        name="themeSelector"
+                        key={`${_theme.id}_${disabled}_${isChecked}`}
```

Comment:
Without it, the radio button is not rendered if I want to unchecked it programmatically (disabled or remove the used theme)

Reply:
I see, so we're relying on the key changing to fully remount it and have `defaultChecked` take effect again. I wasn't expecting that at all, so could you comment it?

The pattern I would expect here is to set the `checked` prop instead of `defaultChecked`, so the component can be controlled programatically without the `key` magic.

Reply:
Oh my last comment disappeared. I used checked in this commit [`9f6ae52` (#12576)](https://github.com/matrix-org/matrix-react-sdk/pull/12576/commits/9f6ae5265a6b4c29d76e3b79d878dd2a9dc0b6e7)